dvc/repo/data.py

```diff
@@ -10,124 +9,98 @@
-
-    if not old and not new:
-        return d
-    if not new:
-        return {"deleted": [old_root], **d}
-    if not old:
-        return {"added": [new_root], **d}
-    if old != new:
-        return {"modified": [new_root], **d}
-    return {"unchanged": [new_root], **d}
-
-
-def _granular_diff(
-    root: str,
-    old_obj: Optional["HashFile"],
-    new_obj: Optional["HashFile"],
-    cache: "HashFileDB",
-) -> Dict[str, List[str]]:
-    from dvc_data.hashfile.diff import ROOT
-    from dvc_data.hashfile.diff import diff as odiff
+    if typ == DELETE:
+        return "deleted"
 
-    def path_join(root: str, *paths: str, isdir: bool = False) -> str:
-        if not isdir and paths == ROOT:
-            return root
-        return os.path.sep.join([root, *paths])
+    return typ
 
-    diff_data = odiff(old_obj, new_obj, cache)
 
-    output: Dict[str, List[str]] = defaultdict(list)
-    for state in ("added", "deleted", "modified", "unchanged"):
-        items = getattr(diff_data, state)
-        for item in items:  # pylint: disable=not-an-iterable
-            entry = item.old if state == "deleted" else item.new
-            isdir = entry.oid.isdir if entry.oid else False
+def _adapt_path(change):
+    isdir = False
+    if change.new and change.new.meta:
+        isdir = change.new.meta.isdir
+    elif change.old and change.old.meta:
+        isdir = change.old.meta.isdir
+    key = change.key
+    if isdir:
+        key = (*key, "")
+    return os.path.sep.join(key)
 
-            path = path_join(root, *entry.key, isdir=isdir)
-            output[state].append(path)
-            if not item.old.in_cache and state != "added":
-                output["not_in_cache"].append(path)
-    return dict(output)
 
+def _diff(
+    old: "DataIndex",
+    new: "DataIndex",
+    *,
+    granular: Optional[bool] = False,
+    with_missing: Optional[bool] = False,
+) -> Dict[str, List[str]]:
+    from dvc_data.index.diff import UNCHANGED, UNKNOWN, diff
+
+    ret: Dict[str, List[str]] = {}
+
+    def _add_change(typ, change):
+        typ = _adapt_typ(typ)
+        if typ not in ret:
+            ret[typ] = []
+
+        ret[typ].append(_adapt_path(change))
+
+    for change in diff(
+        old,
+        new,
+        with_unchanged=True,
+        shallow=not granular,
+        hash_only=True,
+        with_unknown=True,
+    ):
+        if (
+            change.typ == UNCHANGED
+            and change.old
+            and change.new
+            and not change.old.hash_info
+            and not change.new.hash_info
+        ):
+            # NOTE: emulating previous behaviour
+            continue
 
-def _get_obj_items(root: str, obj: Optional["HashFile"]) -> List[str]:
-    if not obj:
-        return []
+        if change.typ == UNKNOWN and not change.new:
+            # NOTE: emulating previous behaviour
+            continue
 
-    from dvc_data.hashfile.tree import Tree
+        if (
+            with_missing
+            and change.old
```

Comment:
I remember there were some discussions around `--no-commit` but could not find it.
The type hints do allow `change.old.hash_info` to not exist, but mypy seems okay with this for some reason. Anyway, I'll try to add support:
```suggestion
            and change.old
            and change.old.hash_info
```

(If you don't have md5 value set, `dvc data status` will fail at the moment which it should not.)